File: 0101-0200/0109.Convert_Sorted_List_to_Binary_Search_Tree.cpp
//2ms

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    void dfs(TreeNode* h, ListNode* &head) {
        if (!h) return;
        dfs(h->left, head);
        h->val = head->val;
        head = head->next;
        dfs(h->right, head);
    }

public:
    TreeNode* sortedListToBST(ListNode* head) {
        if (head == nullptr) return nullptr;

        // Step 1: Count the number of nodes
        int n = 0;
        ListNode* temp = head;
        while (temp) {
            n++;
            temp = temp->next;
        }

        // Step 2: Create an empty tree with the required structure
        TreeNode* root = new TreeNode();
        queue<TreeNode*> q;
        q.push(root);
        n--;

        while (n > 0) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode* node = q.front();
                q.pop();
                if (n > 0) {
                    node->left = new TreeNode();
                    q.push(node->left);
                    n--;
                }
                if (n > 0) {
                    node->right = new TreeNode();
                    q.push(node->right);
                    n--;
                }
            }
        }

        // Step 3: Assign values to the tree nodes
        dfs(root, head);

        return root;
    }
};

File: 0101-0200/0109.Convert_Sorted_List_to_Binary_Search_Tree.js
//1ms


/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {ListNode} head
 * @return {TreeNode}
 */
var sortedListToBST = function(head) {
    if (!head) return null;
    
    let slow = head, fast = head, prev = null;
    
    while (fast && fast.next) {
        prev = slow;
        slow = slow.next;
        fast = fast.next.next;
    }
    
    if (prev) {
        prev.next = null;
    } else {
        head = null;
    }
    
    const root = new TreeNode(slow.val);
    root.left = sortedListToBST(head);
    root.right = sortedListToBST(slow.next);
    
    return root;
};
